### DISCUSSION TITLE
Checking if external ID exists in mutation with Federation

### DISCUSSION THREAD

**Discussion by tmCision** [Netflix/dgs-framework#1158] [Q&A] [answered]
Hello,
I am going over the Federation code example from https://netflix.github.io/dgs/federation/ and I wanted to expand it so to include adding a new review for a show.  I figured out how to add a new review but what I don't know how to do is to check if the associated show actually exists.  Is it possible to retrieve the external show item (I only need the ID) from within the review data fetcher?
Thank you

**Comment by srinivasankavitha** [comment, marked as answer]
Could you please share your code, so it's easier to see what you are trying to achieve?

**Reply by tmCision** [reply]
Sure 
Here is the schema for the Show:
```
type Query {
    shows(titleFilter: String): [Show]
}

type Show @key(fields: "id") {
    id: ID
    title: String
    releaseYear: Int
}
```
There is a show graphql service that provides this information. 

There is a separate service that provides reviews for those shows.
Here is the schema for review
```
type Show @key(fields: "id") @extends {
  id: ID @external
  reviews: [Review]
}

type Review @key(fields: "id") {
  id: ID
  starRating: Int
  body: String
}

type Query {
  reviews: [Review]
}

input ReviewInput {
  id: ID
  starRating: Int
  body: String
  showID: ID
}
type Mutation {
  createReview(input: ReviewInput): Review
}
```

In the Review java project I have this DgsMutation method for the createReview
```
    @DgsMutation
    public Review createReview(@InputArgument ReviewInput input) {
      
        String showID = input.getShowID();
        if (showID != null) {
                 //This is where I want to check if the show actually exists
        }
        Review review = new Review( input.getId(), input.getStarRating(),  input.getBody());
        reviewMap.put(review.getId(), review);
        return review;
      }
```
From the Apollo Gateway I can get a list of shows and their reviews.  My understanding is that the gateway first calls the show service, gets a list of showIDs and than passes that to the review service to get the reviews.  What I am wondering is if there is a way either inside the review service or for the gateway to call the show service to see if the show ID is valid without having to explicitly add the client code.

**Reply by srinivasankavitha** [reply]
Firstly, the show Ids will be populated and passed on to the reviews service only when there is a query that requires the Show type as the return type. In your example, createReview is not going to require the gateway to talk to the Shows service. Rather the showId needs to be provided as part of input to the `createReview` query.

In the case where a `Show` type is returned, the Show service should be taking care of the necessary validation and will return Show objects only for valid showIds.

In the case of the `createReview` query, the gateway will just talk directly to your reviews service and therefore, your service will need to validate that the showId exists.

Hope that clarifies things a bit better.

**Reply by tmCision** [reply]
Thank you for the quick reply.  That is what I was afraid the answer was going to be but wanted to make sure.
